query-parser/src/parse.rs

```diff
@@ -284,6 +290,71 @@ def_parser!(Query, aggregate, Aggregate, {
          })
 });
 
+def_parser!(Query, pull_concrete_attribute, PullAttributeSpec, {
+    forward_keyword().map(|k|
+        PullAttributeSpec::Attribute(
+            PullConcreteAttribute::Ident(
+                ::std::rc::Rc::new(k.clone()))))
+});
+
+def_parser!(Query, pull_wildcard_attribute, PullAttributeSpec, {
+    Query::wildcard().map(|_| PullAttributeSpec::Wildcard)
+});
+
+def_parser!(Query, pull_attribute, PullAttributeSpec, {
+    choice([
+        try(Query::pull_concrete_attribute()),
+        try(Query::pull_wildcard_attribute()),
+        // TODO: reversed keywords, entids (with aliases, presumably…).
+    ])
+});
+
+// A wildcard can appear only once.
+// If a wildcard appears, only map expressions can be present.
+fn validate_attributes<T, E>(attrs: Vec<PullAttributeSpec>) -> std::result::Result<Vec<PullAttributeSpec>, combine::primitives::Error<T, E>> {
```

Comment:
This kind of validation occurs a few places, and I'd like to see it duplicated in the code that processes the data structures.  To that end, could we make this `(attrs: I) -> Option<String> where I: IntoIter<Item=&'a ...>` so that we don't refer to `combine` in the validation and so that we can just check validity by reference?